Project-Aurora/Project-Aurora/Devices/Uniwill/KeyboardDevice.cs

```diff
@@ -0,0 +1,346 @@
+                           
+                    
+                    isInitialized = false;
+                    return false;
+                    // Mark Initialized = TRUE
+                   
+                    }
+                    catch (Exception ex)
+                    {
+                        Global.logger.Error("keyboard device, Exception! Message:" + ex);
+                    }
+                    //try
+                    //{
+                    //    // Initialize  
+                       
+
+
+                    //    // Mark Initialized = TRUE
+                    //    isInitialized = true;
+                    //    return true;
+                    //}
+                    //catch (Exception ex)
+                    //{
+                    //    Global.logger.Error("keyboard device, Exception! Message:" + ex);
+                    //}
+
+
+
+                   // Mark Initialized = FALSE
+                   isInitialized = false;
+                    return false;
+                }
+ 
+
+            return isInitialized;
+
+            }
+
+            // Handle Logon Event
+            //void SystemEvents_SessionSwitch(object sender, SessionSwitchEventArgs e)
+            //{
+            //    if (this.IsInitialized() && e.Reason.Equals(SessionSwitchReason.SessionUnlock))
+            //    { // Only Update when Logged In
+            //        this.SendColorsToKeyboard(true);
+            //    }
+            //}
+
+            public void Shutdown()
+            {
+                if (this.IsInitialized())
+                {
+                    if(CheckGCPower())
+                     keyboard?.release();
+
+                    bRefreshOnce = true;
+                    isInitialized = false;
+                   
+                }
+            }
+
+            public void Reset()
+            {
+                if (this.IsInitialized())
+                {
+                 if(CheckGCPower())
+                    keyboard?.release();
+
+                 bRefreshOnce = true;
+                 isInitialized = false;
+                
+                }
+                   
+            }
+
+            public bool Reconnect()
+            {
+                throw new NotImplementedException();
+            }
+
+            public bool IsInitialized()
+            {
+        
+               return isInitialized;
+            }
+
+            public bool IsConnected()
+            {
+                 return isInitialized;
+           }
+       
+        public bool UpdateDevice(Dictionary<DeviceKeys, Color> keyColors, DoWorkEventArgs e, bool forced = false) // Is this necessary?
+        {
+            if (e.Cancel) return false;
+
+            bool update_result = false;
+
+            watch.Restart();
+
+
+            //Alpha necessary for Global Brightness modifier
+            List<DeviceKeys> KeysList = new List<DeviceKeys>();
+
+            foreach (var key in keyColors.Keys) KeysList.Add(key);
```

Comment:
why do you loop twice here? 
if you are just changing the brightness of keyColors you should instead do something like
```cs
var adjustedColors = keyColors.Keys.Select(kc => AdjustBrightness(kc));

keyboard.SetEffect(adjustedColors);
```

Reply:
Thanks for your suggestion. I will change it.